xds/internal/balancer/outlierdetection/balancer.go

```diff
@@ -105,3 +135,743 @@ func (bb) ParseConfig(s json.RawMessage) (serviceconfig.LoadBalancingConfig, err
+	logger *grpclog.PrefixLogger
+
+	// childMu protects child and also updates to the child (to uphold the
+	// balancer.Balancer API guarantee of synchronous calls). It also protects
+	// against run() reading that the child is not nil for SubConn updates, and
+	// then UpdateClientConnState or Close writing to the the child.
+	childMu sync.Mutex
+	child   balancer.Balancer
+
+	// mu guards access to the following fields. It also helps to synchronize
+	// behaviors of the following events: config updates, firing of the interval
+	// timer, SubConn State updates, SubConn address updates, and child state
+	// updates.
+	//
+	// For example, when we receive a config update in the middle of the
+	// interval timer algorithm, which uses knobs present in the config, the
+	// balancer will wait for the interval timer algorithm to finish before
+	// persisting the new configuration.
+	//
+	// Another example would be the updating of the addrs map, such as from a
+	// SubConn address update in the middle of the interval timer algorithm
+	// which uses addrs. This balancer waits for the interval timer algorithm to
+	// finish before making the update to the addrs map.
+	mu                   sync.Mutex
+	addrs                map[string]*addressInfo
+	cfg                  *LBConfig
+	scWrappers           map[balancer.SubConn]*subConnWrapper
+	timerStartTime       time.Time
+	intervalTimer        *time.Timer
+	inhibitPickerUpdates bool
+	numAddrsEjected      int // For fast calculations of percentage of addrs ejected
+
+	scUpdateCh     *buffer.Unbounded
+	pickerUpdateCh *buffer.Unbounded
+}
+
+// noopConfig returns whether this balancer is configured with a logical no-op
+// configuration or not.
+func (b *outlierDetectionBalancer) noopConfig() bool {
+	return b.cfg.SuccessRateEjection == nil && b.cfg.FailurePercentageEjection == nil
+}
+
+func (b *outlierDetectionBalancer) UpdateClientConnState(s balancer.ClientConnState) error {
+	lbCfg, ok := s.BalancerConfig.(*LBConfig)
+	if !ok {
+		b.logger.Errorf("received config with unexpected type %T: %v", s.BalancerConfig, s.BalancerConfig)
+		return balancer.ErrBadResolverState
+	}
+
+	// Reject whole config if child policy doesn't exist, don't persist it for
+	// later.
+	bb := balancer.Get(lbCfg.ChildPolicy.Name)
+	if bb == nil {
+		return fmt.Errorf("outlier detection: child balancer %q not registered", lbCfg.ChildPolicy.Name)
+	}
+
+	if b.child == nil || b.cfg.ChildPolicy.Name != lbCfg.ChildPolicy.Name {
+		b.childMu.Lock()
+		if b.child != nil {
+			b.child.Close()
+		}
+		b.child = bb.Build(b, b.bOpts)
+		b.childMu.Unlock()
+	}
+
+	b.mu.Lock()
+	b.cfg = lbCfg
+
+	addrs := make(map[string]bool, len(s.ResolverState.Addresses))
+	for _, addr := range s.ResolverState.Addresses {
+		addrs[addr.Addr] = true
+		if _, ok := b.addrs[addr.Addr]; !ok {
+			b.addrs[addr.Addr] = newAddressInfo()
+		}
+	}
+	for addr := range b.addrs {
+		if !addrs[addr] {
+			delete(b.addrs, addr)
+		}
+	}
+
+	var interval time.Duration
```

Comment:
According to the gRFC, this block of code should be done only if the config specifies at least one of the ejection algorithms. https://screenshot.googleplex.com/326HFc3hWjbFuPY.

You could probably simplify this code by adding a couple of helpers, one for the things to be done when a no-op config is specified, and one for the things to be done when a non-noop config is specified.